video/audio_channel.h

```diff
@@ -129,85 +135,104 @@ uint8_t AudioChannel::getStatus() {
 	return status;
 }
 
-std::unique_ptr<fabgl::WaveformGenerator> AudioChannel::getSampleWaveform(uint16_t sampleId, std::shared_ptr<AudioChannel> channelRef) {
+std::shared_ptr<fabgl::WaveformGenerator> AudioChannel::getSampleWaveform(uint16_t sampleId, std::shared_ptr<AudioChannel> channelRef) {
 	if (samples.find(sampleId) != samples.end()) {
 		auto sample = samples.at(sampleId);
+		// if (sample->channels.find(_channel) != sample->channels.end()) {
+		// 	// this channel is already playing this sample, so do nothing
+		// 	debug_log("AudioChannel: already playing sample %d on channel %d\n\r", sampleId, channel());
+		// 	return nullptr;
+		// }
+
+		// // if we're already a waveform, then swap sample
+		// if (this->_waveformType == AUDIO_WAVE_SAMPLE) {
+		// 	// swap sample
+		// 	_mutex.lock();
+		// 	((EnhancedSamplesGenerator *)this->_waveform.get())->setSample(sample);
+		// 	_mutex.unlock();
+		// 	debug_log("AudioChannel: setSample %d on channel %d\n\r", sampleId, channel());
+		// 	return nullptr;
+		// }
+
+		// TODO remove channel tracking??
 		// remove this channel from other samples
-		for (auto samplePair : samples) {
-			if (samplePair.second) {
-				samplePair.second->channels.erase(_channel);
-			}
-		}
-		sample->channels[_channel] = channelRef;
-
-		return make_unique_psram<EnhancedSamplesGenerator>(sample);
+		// for (auto samplePair : samples) {
+		// 	if (samplePair.second) {
+		// 		samplePair.second->channels.erase(_channel);
+		// 	}
+		// }
+		// sample->channels[_channel] = channelRef;
+
+		return make_shared_psram<EnhancedSamplesGenerator>(sample);
 	}
 	return nullptr;
 }
 
 void AudioChannel::setWaveform(int8_t waveformType, std::shared_ptr<AudioChannel> channelRef, uint16_t sampleId) {
-	std::unique_ptr<fabgl::WaveformGenerator> newWaveform = nullptr;
+	std::shared_ptr<fabgl::WaveformGenerator> newWaveform = nullptr;
```

Comment:
I changed the waveform in this function, and in the class to be a shared_ptr rather than a unique_ptr.  this was done in part because I'd changed the underlying API in vdp-gl to use a shared_ptr.  the main point of this was to try to ensure the object would be retained until the last uses of it had gone away

as it goes, I stepped back from that change to the vdp-gl API, so perhaps this class should change back to using a unique_ptr too?

Reply:
left as shared_ptr cos we're no longer creating our waveform generator objects in psram, and making unique_ptr's in C++11 syntax is awkward